### ISSUE TITLE
False positive 'option-map-unit-fn' for std::process::exit

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#5180]
```rust
warning: called `map(f)` on an Option value where `f` is a unit function
   --> src/bin/rcc.rs:124:9
    |
124 |         unsafe { rccjit.run_main() }.map(std::process::exit);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
    |         |
    |         help: try this: `if let Some(_) = unsafe { rccjit.run_main() } { std::process::exit(_) }`
```

This suggestion makes no sense: you can't pass `_` to a function. Furthermore, `exit()` is not a unit function, it takes `i32`.

`cargo clippy -V`: `clippy 0.0.212 (e8642c7 2020-01-06)` (but also happens in stable)

**Comment by flip1995** [comment]
> Furthermore, exit() is not a unit function, it takes i32.

The unit functions refers to the return value, not the param. `!` is treated like the unit type in this lint apparently. The suggestion uses `_` as a placeholder. Maybe we could use `a` or something similar as a placeholder instead.

**Comment by jyn514** [comment]
> The unit functions refers to the return value, not the param. 

I see, that makes more sense. That wasn't very clear from the suggestion, maybe it could say 'f is a function returning unit type' instead?

**Comment by jyn514** [comment]
That PR (#5292) didn't address my main concern, which is that 'unit function' is not very clear. Can this be reopened?

**Comment by flip1995** [comment]
#5290 addressed this comment https://github.com/rust-lang/rust-clippy/issues/5180#issuecomment-586789264

With #5292, this issue should be resolved completely, right?

**Comment by jyn514** [comment]
Oops, I didn't notice that #5290 had been merged. In that case I'm happy with the fix :)
